langchain4j/src/main/java/dev/langchain4j/service/DefaultAiServices.java

```diff
@@ -181,11 +184,20 @@ public Object invoke(Object proxy, Method method, Object[] args) throws Exceptio
                                         toolExecutionRequest,
                                         toolExecutionResult
                                 );
+
                                 if (context.hasChatMemory()) {
                                     context.chatMemory(memoryId).add(toolExecutionResultMessage);
                                 } else {
                                     messages.add(toolExecutionResultMessage);
                                 }
+                                if (method.isAnnotationPresent(ReturnAsOutput.class)) {
+                                    if (context.hasChatMemory()) {
+                                        context.chatMemory(memoryId).add(AiMessage.aiMessage(toolExecutionResult));
```

Comment:
Many LLM providers (such as OpenAI) will fail if `AiMessage` with tool execution request is not followed by a tool retult message

Reply:
thanks for you review , this is maybe not a problem ,make tool method return directly also add “ toolExecutionResultMessage” ,see my comment in code lines

Reply:
My bad, I misunderstood the change. But then why `AiMessage.aiMessage(toolExecutionResult)` is added to the memory if `toolExecutionResultMessage` has been already added?

Reply:
I will give you a example chat history to illustrate this:
```Json 
[
        {
            "role": "user",
            "content": "how about the weather in Shanghai？"
        },
        {
            "role": "assistant",
            "content": null,
            "function_call": {
                "name": "get_current_weather",
                "arguments": "{ \"location\": \"Shanghai\"}"
            }
        },
        {----here is toolExecutionResultMessage,without this ,some LLM providers (such as OpenAI) will fail 
            "role": "function",
            "name": "get_current_weather",
            "content": "{\"temperature\": \"25\", \"unit\": \"degree\", \"description\": \"sunny\"}"
        },
        {----here is toolExecutionResult using Tool function return value(nomally,we need request LLM to generate final answer ,here we use Tool function return value as final answer to reduce LLM requests )
            "role": "assistant",
            "content": "It's 25 degrees Celsius and sunny in Shanghai todad"
        }]
```

Reply:
Ok I see.

What is the use case for this? Do you plan to use this with chat memory? Is this for a "data extraction" use case?

Reply:
> Ok I see.
> 
> What is the use case for this? Do you plan to use this with chat memory? Is this for a "data extraction" use case?
I update the example chat history given to you before,this case is only for function call, and in function call must use chat memory. 
I think this feature could make function call more reliable，sometimes we need LLM call funtion and output strict json string， but llm may not ,if we can use tool function return value as output ,we can control a predictable output

Reply:
> this case is only for function call

If this change is to use function calls to extract structured data, I would say that this is a misuse of the feature.

> and in function call must use chat memory.

Memory is not needed any more to use tools.

> I think this feature can make function call more reliable and more Fault-tolerant，sometimes we need LLM call funtion and output strict json string， but llm may not ,if we can use tool function return value as output ,We can expect to get a definite output(cuz we can control tool function return object or strict json )

Agree, we have an [issue](https://github.com/langchain4j/langchain4j/issues/859) to address this. Soon one could use function calls to extract structured data. But for that you don't need to define a `@Tool` method. You could just define a POJO as a return type.